File: swift/Sources/Fory/TypeResolver.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

private struct TypeNameKey: Hashable {
    let namespace: String
    let typeName: String
}

final class TypeResolver {
    private let trackRef: Bool

    private var bySwiftType: [ObjectIdentifier: TypeInfo] = [:]
    private var byUserTypeID: [UInt32: TypeInfo] = [:]
    private var byTypeName: [TypeNameKey: TypeInfo] = [:]
    private var builtinTypeInfoByID: [TypeId: TypeInfo] = [:]
    private var compatibleTypeInfoByHeader: [UInt64: TypeInfo] = [:]

    init(trackRef: Bool = false) {
        self.trackRef = trackRef
    }

    func register<T: Serializer>(_ type: T.Type, id: UInt32) {
        do {
            try registerByID(type, id: id)
        } catch {
            preconditionFailure("conflicting registration for \(type): \(error)")
        }
    }

    private func registerByID<T: Serializer>(_ type: T.Type, id: UInt32) throws {
        let swiftTypeID = ObjectIdentifier(type)
        try validateIDRegistration(key: swiftTypeID, type: type, id: id)

        let typeInfo = try TypeInfo(
            swiftTypeID: swiftTypeID,
            typeID: T.staticTypeId,
            userTypeID: id,
            registerByName: false,
            namespace: MetaString.empty(specialChar1: ".", specialChar2: "_"),
            typeName: MetaString.empty(specialChar1: "$", specialChar2: "_"),
            fields: T.foryFieldsInfo(trackRef: trackRef),
            reader: { context in
                try T.foryRead(context, refMode: .none, readTypeInfo: false)
            },
            compatibleReader: { context, remoteTypeInfo in
                try T.foryReadCompatibleData(context, remoteTypeInfo: remoteTypeInfo)
            }
        )

        if let existing = bySwiftType[swiftTypeID],
           existing.matches(
               typeID: T.staticTypeId,
               userTypeID: id,
               registerByName: false,
               namespace: "",
               typeName: ""
           ) {
            return
        }

        try store(typeInfo, for: swiftTypeID, userTypeID: id)
    }

    func register<T: Serializer>(_ type: T.Type, namespace: String, typeName: String) throws {
        let namespaceMeta = try MetaStringEncoder.namespace.encode(
            namespace,
            allowedEncodings: namespaceMetaStringEncodings
        )
        let typeNameMeta = try MetaStringEncoder.typeName.encode(
            typeName,
            allowedEncodings: typeNameMetaStringEncodings
        )
        let swiftTypeID = ObjectIdentifier(type)
        try validateNameRegistration(
            key: swiftTypeID,
            type: type,
            namespace: namespace,
            typeName: typeName
        )

        let typeInfo = try TypeInfo(
            swiftTypeID: swiftTypeID,
            typeID: T.staticTypeId,
            userTypeID: nil,
            registerByName: true,
            namespace: namespaceMeta,
            typeName: typeNameMeta,
            fields: T.foryFieldsInfo(trackRef: trackRef),
            reader: { context in
                try T.foryRead(context, refMode: .none, readTypeInfo: false)
            },
            compatibleReader: { context, remoteTypeInfo in
                try T.foryReadCompatibleData(context, remoteTypeInfo: remoteTypeInfo)
            }
        )

        if let existing = bySwiftType[swiftTypeID],
           existing.matches(
               typeID: T.staticTypeId,
               userTypeID: nil,
               registerByName: true,
               namespace: namespace,
               typeName: typeName
           ) {
            return
        }

        try store(typeInfo, for: swiftTypeID, typeNameKey: TypeNameKey(namespace: namespace, typeName: typeName))
    }

    func register<T: Serializer>(_ type: T.Type, name: String) throws {
        let parts = name.components(separatedBy: ".")
        if parts.count <= 1 {
            try register(type, namespace: "", typeName: name)
            return
        }

        let resolvedTypeName = parts[parts.count - 1]
        let resolvedNamespace = parts.dropLast().joined(separator: ".")
        try register(type, namespace: resolvedNamespace, typeName: resolvedTypeName)
    }

    func requireTypeInfo<T: Serializer>(for type: T.Type) throws -> TypeInfo {
        if let info = bySwiftType[ObjectIdentifier(type)] {
            return info
        }
        throw ForyError.typeNotRegistered("\(type) is not registered")
    }

    @inline(__always)
    func compatibleTypeInfo(forHeader header: UInt64) -> TypeInfo? {
        compatibleTypeInfoByHeader[header]
    }

    @inline(__always)
    func cacheCompatibleTypeInfo(_ typeMeta: TypeMeta, forHeader header: UInt64) throws -> TypeInfo {
        if let cached = compatibleTypeInfoByHeader[header] {
            return cached
        }
        let localTypeInfo = try requireCompatibleTypeInfo(for: typeMeta)
        if header == localTypeInfo.typeDefHeader {
            compatibleTypeInfoByHeader[header] = localTypeInfo
            return localTypeInfo
        }
        let canonicalTypeMeta: TypeMeta
        if let localTypeMeta = localTypeInfo.typeMeta,
           let remapped = try? typeMeta.assigningFieldIDs(from: localTypeMeta) {
            canonicalTypeMeta = remapped
        } else {
            canonicalTypeMeta = typeMeta
        }
        let compatibleTypeInfo = TypeInfo(dynamic: localTypeInfo, compatibleTypeMeta: canonicalTypeMeta)
        compatibleTypeInfoByHeader[header] = compatibleTypeInfo
        return compatibleTypeInfo
    }

    private func store(
        _ typeInfo: TypeInfo,
        for swiftTypeID: ObjectIdentifier,
        userTypeID: UInt32? = nil,
        typeNameKey: TypeNameKey? = nil
    ) throws {
        bySwiftType[swiftTypeID] = typeInfo
        if let userTypeID {
            byUserTypeID[userTypeID] = typeInfo
        }
        if let typeNameKey {
            byTypeName[typeNameKey] = typeInfo
        }
        if let typeMeta = typeInfo.typeMeta,
           let typeDefHeader = typeInfo.typeDefHeader {
            compatibleTypeInfoByHeader[typeDefHeader] = TypeInfo(
                dynamic: typeInfo,
                compatibleTypeMeta: typeMeta
            )
        }
    }

    @inline(__always)
    func builtinTypeInfo(for typeID: TypeId) -> TypeInfo {
        if let cached = builtinTypeInfoByID[typeID] {
            return cached
        }
        let info = TypeInfo(typeID: typeID)
        builtinTypeInfoByID[typeID] = info
        return info
    }

    @inline(__always)
    func requireTypeInfo(userTypeID: UInt32) throws -> TypeInfo {
        guard let typeInfo = byUserTypeID[userTypeID] else {
            throw ForyError.typeNotRegistered("user_type_id=\(userTypeID)")
        }
        return typeInfo
    }

    @inline(__always)
    func requireTypeInfo(namespace: String, typeName: String) throws -> TypeInfo {
        guard let typeInfo = byTypeName[TypeNameKey(namespace: namespace, typeName: typeName)] else {
            throw ForyError.typeNotRegistered("namespace=\(namespace), type=\(typeName)")
        }
        return typeInfo
    }

    private func validateIDRegistration<T: Serializer>(
        key: ObjectIdentifier,
        type: T.Type,
        id: UInt32
    ) throws {
        if let existing = bySwiftType[key] {
            if existing.registerByName {
                throw ForyError.invalidData(
                    "\(type) was already registered by name, cannot re-register by id"
                )
            }
            if existing.typeID != T.staticTypeId || existing.userTypeID != id {
                let existingID = existing.userTypeID.map { String($0) } ?? "nil"
                throw ForyError.invalidData(
                    "\(type) registration conflict: existing id=\(existingID), new id=\(id)"
                )
            }
        }

        if let existing = byUserTypeID[id], existing.swiftTypeID != key {
            throw ForyError.invalidData("user type id \(id) is already registered by another type")
        }
    }

    private func validateNameRegistration<T: Serializer>(
        key: ObjectIdentifier,
        type: T.Type,
        namespace: String,
        typeName: String
    ) throws {
        if let existing = bySwiftType[key] {
            if !existing.registerByName {
                throw ForyError.invalidData(
                    "\(type) was already registered by id, cannot re-register by name"
                )
            }
            if existing.typeID != T.staticTypeId ||
                existing.namespace.value != namespace ||
                existing.typeName.value != typeName {
                throw ForyError.invalidData(
                    """
                    \(type) registration conflict: existing name=\(existing.namespace.value)::\(existing.typeName.value), \
                    new name=\(namespace)::\(typeName)
                    """
                )
            }
        }

        let nameKey = TypeNameKey(namespace: namespace, typeName: typeName)
        if let existing = byTypeName[nameKey], existing.swiftTypeID != key {
            throw ForyError.invalidData("type name \(namespace)::\(typeName) is already registered by another type")
        }
    }

    @inline(__always)
    private func requireCompatibleTypeInfo(for typeMeta: TypeMeta) throws -> TypeInfo {
        if typeMeta.registerByName {
            guard let typeInfo = byTypeName[TypeNameKey(namespace: typeMeta.namespace.value, typeName: typeMeta.typeName.value)] else {
                throw ForyError.typeNotRegistered(
                    "namespace=\(typeMeta.namespace.value), type=\(typeMeta.typeName.value)"
                )
            }
            return typeInfo
        }
        if let userTypeID = typeMeta.userTypeID {
            guard let typeInfo = byUserTypeID[userTypeID] else {
                throw ForyError.typeNotRegistered("user_type_id=\(userTypeID)")
            }
            return typeInfo
        }
        throw ForyError.invalidData("missing user type id in compatible dynamic type meta")
    }

}

File: javascript/packages/core/lib/typeResolver.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  ForyTypeInfoSymbol,
  WithForyClsInfo,
  Serializer,
  TypeId,
  MaxInt32,
  MinInt32,
  Config,
} from "./type";
import { Gen } from "./gen";
import { Dynamic, Type, TypeInfo } from "./typeInfo";
import { ReadContext, WriteContext } from "./context";
import { Decimal } from "./types/decimal";
import { BFloat16Array } from "./types/bfloat16";
import { BoolArray } from "./types/boolArray";
import { isFloat16Array } from "./types/float16";
import { getUnknownTypeMeta, UnknownStructSerializer } from "./unknownStruct";

export default class TypeResolver {
  readonly trackingRef: boolean;
  private internalSerializer: Serializer[] = new Array(300);
  private customSerializer: Map<number | string, Serializer> = new Map();
  private registrationFrozen = false;

  private writeContext!: WriteContext;
  private readContext!: ReadContext;

  private float64Serializer: null | Serializer = null;
  private float32Serializer: null | Serializer = null;
  private varint32Serializer: null | Serializer = null;
  private varInt64Serializer: null | Serializer = null;
  private int64Serializer: null | Serializer = null;
  private boolSerializer: null | Serializer = null;
  private datetimeSerializer: null | Serializer = null;
  private decimalSerializer: null | Serializer = null;
  private stringSerializer: null | Serializer = null;
  private setSerializer: null | Serializer = null;
  private listSerializer: null | Serializer = null;
  private mapSerializer: null | Serializer = null;
  private uint8ArraySerializer: null | Serializer = null;
  private uint16ArraySerializer: null | Serializer = null;
  private uint32ArraySerializer: null | Serializer = null;
  private uint64ArraySerializer: null | Serializer = null;
  private int8ArraySerializer: null | Serializer = null;
  private int16ArraySerializer: null | Serializer = null;
  private int32ArraySerializer: null | Serializer = null;
  private int64ArraySerializer: null | Serializer = null;
  private boolArraySerializer: null | Serializer = null;
  private float16ArraySerializer: null | Serializer = null;
  private bfloat16ArraySerializer: null | Serializer = null;
  private float32ArraySerializer: null | Serializer = null;
  private float64ArraySerializer: null | Serializer = null;
  private unknownStructSerializer!: UnknownStructSerializer;

  constructor(readonly config: Config) {
    this.trackingRef = config.ref;
  }

  bindContexts(writeContext: WriteContext, readContext: ReadContext) {
    this.writeContext = writeContext;
    this.readContext = readContext;
    this.unknownStructSerializer = new UnknownStructSerializer(this, writeContext, readContext);
  }

  getUnknownStructSerializer(typeMeta?: import("./meta/TypeMeta").TypeMeta, wireTypeId?: number) {
    return typeMeta === undefined
      ? this.unknownStructSerializer
      : this.unknownStructSerializer.createReadSerializer(typeMeta, wireTypeId);
  }

  isCompatible() {
    return this.config.compatible === true;
  }

  computeTypeId(typeInfo: TypeInfo) {
    const internalTypeId = typeInfo.typeId;
    if (internalTypeId !== TypeId.STRUCT && internalTypeId !== TypeId.NAMED_STRUCT) {
      return internalTypeId;
    }
    if (internalTypeId === TypeId.NAMED_STRUCT && this.isCompatible() && typeInfo.evolving) {
      return TypeId.NAMED_COMPATIBLE_STRUCT;
    }
    if (internalTypeId === TypeId.STRUCT && this.isCompatible() && typeInfo.evolving) {
      return TypeId.COMPATIBLE_STRUCT;
    }
    return internalTypeId;
  }

  isMonomorphic(typeInfo: TypeInfo, dynamic: Dynamic = Dynamic.AUTO) {
    switch (dynamic) {
      case Dynamic.TRUE:
        return false;
      case Dynamic.FALSE:
        return true;
      default:
        if (TypeId.enumType(typeInfo.typeId)) {
          return true;
        }
        if (
          typeInfo.typeId === TypeId.UNION ||
          typeInfo.typeId === TypeId.TYPED_UNION ||
          typeInfo.typeId === TypeId.NAMED_UNION
        ) {
          return true;
        }
        if (this.isCompatible()) {
          return !TypeId.userDefinedType(typeInfo.typeId) && typeInfo.typeId !== TypeId.UNKNOWN;
        }
        return typeInfo.typeId !== TypeId.UNKNOWN;
    }
  }

  private makeUserTypeKey(userTypeId: number) {
    return `u:${userTypeId}`;
  }

  private initInternalSerializer() {
    const generateInternalSerializer = (typeInfo: TypeInfo) => {
      return new Gen(this).generateSerializer(typeInfo);
    };
    generateInternalSerializer(Type.string());
    generateInternalSerializer(new TypeInfo(TypeId.ENUM));
    generateInternalSerializer(new TypeInfo(TypeId.NAMED_ENUM));
    generateInternalSerializer(Type.any());
    generateInternalSerializer(Type.list(Type.any()));
    generateInternalSerializer(Type.map(Type.any(), Type.any()));
    generateInternalSerializer(Type.bool());
    generateInternalSerializer(Type.int8());
    generateInternalSerializer(Type.int16());
    generateInternalSerializer(Type.int32({ encoding: "fixed" }));
    generateInternalSerializer(Type.int32());
    generateInternalSerializer(Type.uint32({ encoding: "fixed" }));
    generateInternalSerializer(Type.uint64({ encoding: "fixed" }));
    generateInternalSerializer(Type.int64({ encoding: "fixed" }));
    generateInternalSerializer(Type.int64());
    generateInternalSerializer(Type.uint8());
    generateInternalSerializer(Type.uint16());
    generateInternalSerializer(Type.uint32());
    generateInternalSerializer(Type.uint64());
    generateInternalSerializer(Type.uint64({ encoding: "tagged" }));
    generateInternalSerializer(Type.int64({ encoding: "tagged" }));
    generateInternalSerializer(Type.float16());
    generateInternalSerializer(Type.bfloat16());
    generateInternalSerializer(Type.float32());
    generateInternalSerializer(Type.float64());
    generateInternalSerializer(Type.timestamp());
    generateInternalSerializer(Type.duration());
    generateInternalSerializer(Type.date());
    generateInternalSerializer(Type.decimal());
    generateInternalSerializer(Type.set(Type.any()));
    generateInternalSerializer(Type.binary());
    generateInternalSerializer(Type.boolArray());
    generateInternalSerializer(Type.uint8Array());
    generateInternalSerializer(Type.int8Array());
    generateInternalSerializer(Type.uint16Array());
    generateInternalSerializer(Type.int16Array());
    generateInternalSerializer(Type.uint32Array());
    generateInternalSerializer(Type.int32Array());
    generateInternalSerializer(Type.uint64Array());
    generateInternalSerializer(Type.int64Array());
    generateInternalSerializer(Type.float16Array());
    generateInternalSerializer(Type.bfloat16Array());
    generateInternalSerializer(Type.float32Array());
    generateInternalSerializer(Type.float64Array());

    this.float64Serializer = this.getSerializerById(TypeId.FLOAT64);
    this.float32Serializer = this.getSerializerById(TypeId.FLOAT32);
    this.varint32Serializer = this.getSerializerById(TypeId.VARINT32);
    this.varInt64Serializer = this.getSerializerById(TypeId.VARINT64);
    this.int64Serializer = this.getSerializerById(TypeId.INT64);
    this.boolSerializer = this.getSerializerById(TypeId.BOOL);
    this.datetimeSerializer = this.getSerializerById(TypeId.TIMESTAMP);
    this.decimalSerializer = this.getSerializerById(TypeId.DECIMAL);
    this.stringSerializer = this.getSerializerById(TypeId.STRING);
    this.setSerializer = this.getSerializerById(TypeId.SET);
    this.listSerializer = this.getSerializerById(TypeId.LIST);
    this.mapSerializer = this.getSerializerById(TypeId.MAP);
    this.uint8ArraySerializer = this.getSerializerById(TypeId.UINT8_ARRAY);
    this.uint16ArraySerializer = this.getSerializerById(TypeId.UINT16_ARRAY);
    this.uint32ArraySerializer = this.getSerializerById(TypeId.UINT32_ARRAY);
    this.uint64ArraySerializer = this.getSerializerById(TypeId.UINT64_ARRAY);
    this.int8ArraySerializer = this.getSerializerById(TypeId.INT8_ARRAY);
    this.int16ArraySerializer = this.getSerializerById(TypeId.INT16_ARRAY);
    this.int32ArraySerializer = this.getSerializerById(TypeId.INT32_ARRAY);
    this.int64ArraySerializer = this.getSerializerById(TypeId.INT64_ARRAY);
    this.boolArraySerializer = this.getSerializerById(TypeId.BOOL_ARRAY);
    this.float16ArraySerializer = this.getSerializerById(TypeId.FLOAT16_ARRAY);
    this.bfloat16ArraySerializer = this.getSerializerById(TypeId.BFLOAT16_ARRAY);
    this.float32ArraySerializer = this.getSerializerById(TypeId.FLOAT32_ARRAY);
    this.float64ArraySerializer = this.getSerializerById(TypeId.FLOAT64_ARRAY);
  }

  init() {
    this.initInternalSerializer();
  }

  freezeRegistration() {
    if (!this.registrationFrozen) {
      this.registrationFrozen = true;
    }
  }

  ensureRegistrationOpen() {
    if (this.registrationFrozen) {
      throw new Error("types and serializers must be registered before the first root operation");
    }
  }

  commitGeneratedSerializers(entries: readonly { typeInfo: TypeInfo; serializer: Serializer }[]) {
    this.ensureRegistrationOpen();
    const publications = entries.map((entry) => {
      if (!entry.serializer._initialized) {
        throw new Error("generated serializer graph is incomplete");
      }
      const typeId = this.computeTypeId(entry.typeInfo);
      let internalTypeId: number | undefined;
      let customTypeKey: number | string | undefined;
      if (TypeId.isNamedType(typeId)) {
        customTypeKey = entry.typeInfo.named!;
      } else if (TypeId.needsUserTypeId(typeId) && entry.typeInfo.userTypeId !== -1) {
        customTypeKey = this.makeUserTypeKey(entry.typeInfo.userTypeId);
      } else if (typeId <= 0xff) {
        internalTypeId = typeId;
      } else {
        customTypeKey = typeId;
      }
      const existingSerializer =
        internalTypeId === undefined
          ? this.customSerializer.get(customTypeKey!)
          : this.internalSerializer[internalTypeId];
      return {
        entry,
        internalTypeId,
        customTypeKey,
        existingSerializer,
      };
    });
    for (const publication of publications) {
      if (publication.existingSerializer !== undefined) {
        continue;
      }
      if (publication.internalTypeId !== undefined) {
        this.internalSerializer[publication.internalTypeId] = publication.entry.serializer;
      } else {
        this.customSerializer.set(publication.customTypeKey!, publication.entry.serializer);
      }
    }
  }

  generateReadSerializer(typeInfo: TypeInfo) {
    return new Gen(this, { creator: typeInfo.options?.creator }).reGenerateSerializer(typeInfo);
  }

  getSerializerByTypeInfo(typeInfo: TypeInfo) {
    const typeId = this.computeTypeId(typeInfo);
    if (TypeId.isNamedType(typeId)) {
      return this.customSerializer.get(typeInfo.named!);
    }
    return this.getSerializerById(typeId, typeInfo.userTypeId);
  }

  getSerializerById(id: number, userTypeId?: number) {
    if (TypeId.needsUserTypeId(id) && userTypeId !== undefined && userTypeId !== -1) {
      return this.customSerializer.get(this.makeUserTypeKey(userTypeId))!;
    }
    if (id <= 0xff) {
      return this.internalSerializer[id]!;
    }
    return this.customSerializer.get(id)!;
  }

  getSerializerByName(typeIdOrName: number | string) {
    return this.customSerializer.get(typeIdOrName);
  }

  getSerializerByData(v: any) {
    if (v === null || v === undefined) {
      return null;
    }
    if (typeof v === "number") {
      if (Number.isInteger(v)) {
        if (v > MaxInt32 || v < MinInt32) {
          return this.varInt64Serializer;
        }
        return this.varint32Serializer;
      }
      if (v > MaxInt32 || v < MinInt32) {
        return this.float64Serializer;
      }
      return this.float32Serializer;
    }

    if (typeof v === "bigint") {
      return this.varInt64Serializer;
    }

    if (typeof v === "string") {
      return this.stringSerializer;
    }

    if (v instanceof Decimal) {
      return this.decimalSerializer;
    }

    if (v instanceof BoolArray) {
      return this.boolArraySerializer;
    }

    if (isFloat16Array(v)) {
      return this.float16ArraySerializer;
    }

    if (v instanceof BFloat16Array) {
      return this.bfloat16ArraySerializer;
    }

    if (v instanceof Uint8Array) {
      return this.uint8ArraySerializer;
    }

    if (v instanceof Uint16Array) {
      return this.uint16ArraySerializer;
    }

    if (v instanceof Uint32Array) {
      return this.uint32ArraySerializer;
    }

    if (v instanceof BigUint64Array) {
      return this.uint64ArraySerializer;
    }

    if (v instanceof Int8Array) {
      return this.int8ArraySerializer;
    }

    if (v instanceof Int16Array) {
      return this.int16ArraySerializer;
    }

    if (v instanceof Int32Array) {
      return this.int32ArraySerializer;
    }

    if (v instanceof BigInt64Array) {
      return this.int64ArraySerializer;
    }

    if (v instanceof Float32Array) {
      return this.float32ArraySerializer;
    }

    if (v instanceof Float64Array) {
      return this.float64ArraySerializer;
    }

    if (Array.isArray(v)) {
      return this.listSerializer;
    }

    if (typeof v === "boolean") {
      return this.boolSerializer;
    }

    if (typeof v === "bigint") {
      return this.int64Serializer;
    }

    if (v instanceof Date) {
      return this.datetimeSerializer;
    }

    if (v instanceof Map) {
      return this.mapSerializer;
    }

    if (v instanceof Set) {
      return this.setSerializer;
    }

    if (getUnknownTypeMeta(v) !== undefined) {
      return this.unknownStructSerializer;
    }

    if (typeof v === "object" && v !== null && ForyTypeInfoSymbol in v) {
      const typeInfo = (v[ForyTypeInfoSymbol] as WithForyClsInfo).structTypeInfo;
      return this.getSerializerByTypeInfo(typeInfo);
    }

    throw new Error(`Failed to detect the Fory type from JavaScript type: ${typeof v}`);
  }
}
